File: prover-android/app/src/main/java/org/nachweis/prover/net/Handoff.kt
package org.nachweis.prover.net

import org.json.JSONObject
import org.nachweis.prover.circuit.Codec

/**
 * Two-device handoff. The investor's browser created the bridge session and signed it with
 * the wallet (EIP-191 address proof); this phone, which holds no Ethereum key, joins THAT
 * session: same bound address and the same 32-byte challenge, so the KB-JWT nonce the relay
 * requests equals the bridge session's nonce, and the proof is posted to that session id.
 *
 * Wire forms (app/src/lib/handoff.ts produces them, companion/src/handoff.ts parses the same):
 *   compact JSON (the QR):  {"v":1,"s":"<session_id>","a":"0x<address>","c":"<challenge_hex>","r":"<verifier_url>","b":"<bridge_url>"}
 *   URI (paste):            nachweis://handoff?v=1&s=...&a=...&c=...&r=...&b=...
 *   bridge body:            GET /sessions/:id/handoff {session_id, bound_address, challenge_hex, nonce, verifier_url, bridge_url, expires_at}
 */
data class Handoff(
    val sessionId: String,
    /** 0x + 40 lowercase hex. */
    val boundAddress: String,
    /** 64 lowercase hex chars, no 0x. */
    val challengeHex: String,
    /** sha256(address20 || challenge32), 64 lowercase hex chars. */
    val nonce: String,
    val verifierUrl: String?,
    val bridgeUrl: String?,
) {
    companion object {
        private val UUID = Regex("^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$")

        /** Accepts the compact JSON, the bridge's handoff body, or the nachweis://handoff URI. */
        fun parse(input: String): Handoff {
            val s = input.trim()
            if (s.startsWith("{")) return fromFields(JSONObject(s).let { o -> o.keys().asSequence().associateWith { k -> if (o.isNull(k)) null else o.get(k).toString() } })
            val prefix = "nachweis://handoff"
            if (s.startsWith(prefix, ignoreCase = true)) {
                val query = s.substringAfter('?', "")
                require(query.isNotEmpty()) { "handoff: URI has no query" }
                val fields = query.split('&').filter { it.isNotEmpty() }.associate { kv ->
                    val i = kv.indexOf('=')
                    val k = if (i < 0) kv else kv.substring(0, i)
                    val v = if (i < 0) "" else java.net.URLDecoder.decode(kv.substring(i + 1), "UTF-8")
                    k to v
                }
                return fromFields(fields)
            }
            throw IllegalArgumentException("handoff: expected a JSON object or a nachweis://handoff?... URI")
        }

        private fun fromFields(f: Map<String, String?>): Handoff {
            val v = f["v"] ?: f["version"]
            require(v == null || v == "1") { "handoff version $v not supported (expected 1)" }
            val session = (f["s"] ?: f["session_id"] ?: "").trim()
            require(UUID.matches(session)) { "handoff: session id is not a UUID" }
            val address = (f["a"] ?: f["bound_address"] ?: "").trim().lowercase()
            require(Regex("^0x[0-9a-f]{40}$").matches(address)) { "handoff: bound address must be 0x + 40 hex chars" }
            val challenge = (f["c"] ?: f["challenge_hex"] ?: "").trim().lowercase().removePrefix("0x")
            require(Regex("^[0-9a-f]{64}$").matches(challenge)) { "handoff: challenge must be 32 bytes of hex" }
            val nonce = Codec.toHex(Codec.sha256(Codec.hex(address) + Codec.hex(challenge)))
            val given = (f["nonce"])?.trim()?.lowercase()?.removePrefix("0x")
            require(given == null || given.isEmpty() || given == nonce) { "handoff: nonce $given is not sha256(address || challenge) $nonce" }
            fun url(u: String?): String? {
                val t = u?.trim()?.trimEnd('/') ?: return null
                if (t.isEmpty() || t == "null") return null
                require(t.startsWith("http://") || t.startsWith("https://")) { "handoff: URL $t is not http(s)" }
                return t
            }
            return Handoff(
                sessionId = session.lowercase(),
                boundAddress = address,
                challengeHex = challenge,
                nonce = nonce,
                verifierUrl = url(f["r"] ?: f["verifier_url"]),
                bridgeUrl = url(f["b"] ?: f["bridge_url"]),
            )
        }
    }
}

File: companion/src/handoff.ts
// Two-device handoff: the browser bound a bridge session to its wallet (EIP-191 address proof);
// this device joins that session and proves. Produced by the web app (app/src/lib/handoff.ts)
// and by the bridge (`GET /sessions/:id/handoff`, field `uri`); the Android app parses the same
// two forms (prover-android Handoff.kt).
//
//   compact JSON (the QR):  {"v":1,"s":"<session_id>","a":"0x<address>","c":"<challenge_hex>","r":"<verifier_url>","b":"<bridge_url>"}
//   URI (copy and paste):   nachweis://handoff?v=1&s=...&a=...&c=...&r=...&b=...
//   bridge JSON:            {session_id, bound_address, challenge_hex, nonce, verifier_url, bridge_url, expires_at}
import { hex, nonceOf, parseAddress } from "./util";

export interface Handoff {
  session_id: string;
  bound_address: string; // 0x + 40 lowercase hex
  challenge_hex: string; // 64 lowercase hex chars, no 0x
  nonce: string; // sha256(address20 || challenge32), 64 lowercase hex chars
  verifier_url?: string;
  bridge_url?: string;
}

const UUID = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

function fromFields(f: Record<string, unknown>): Handoff {
  const v = f.v ?? f.version;
  if (v !== undefined && String(v) !== "1") throw new Error(`handoff version ${String(v)} not supported (expected 1)`);
  const session = String(f.s ?? f.session_id ?? "");
  if (!UUID.test(session)) throw new Error("handoff: session id is not a UUID");
  const addressBytes = parseAddress(String(f.a ?? f.bound_address ?? ""));
  const challengeRaw = String(f.c ?? f.challenge_hex ?? "").toLowerCase().replace(/^0x/, "");
  if (!/^[0-9a-f]{64}$/.test(challengeRaw)) throw new Error("handoff: challenge must be 32 bytes of hex");
  const strip = (u: unknown): string | undefined => {
    if (u === undefined || u === null || u === "") return undefined;
    const s = String(u).replace(/\/+$/, "");
    if (!/^https?:\/\//.test(s)) throw new Error(`handoff: URL ${s} is not http(s)`);
    return s;
  };
  const challenge = Buffer.from(challengeRaw, "hex");
  const nonce = hex(nonceOf(addressBytes, challenge));
  const given = f.nonce !== undefined ? String(f.nonce).toLowerCase().replace(/^0x/, "") : undefined;
  if (given && given !== nonce) throw new Error(`handoff: nonce ${given} is not sha256(address || challenge) ${nonce}`);
  return {
    session_id: session.toLowerCase(),
    bound_address: "0x" + hex(addressBytes),
    challenge_hex: challengeRaw,
    nonce,
    verifier_url: strip(f.r ?? f.verifier_url),
    bridge_url: strip(f.b ?? f.bridge_url),
  };
}

/** Accepts the compact JSON, the bridge's handoff JSON, or the nachweis://handoff URI. */
export function parseHandoff(input: string): Handoff {
  const s = input.trim();
  if (s.startsWith("{")) return fromFields(JSON.parse(s) as Record<string, unknown>);
  const m = /^nachweis:\/\/handoff\/?\?(.*)$/i.exec(s);
  if (m) {
    const q = new URLSearchParams(m[1]);
    const f: Record<string, unknown> = {};
    for (const [k, v] of q) f[k] = v;
    return fromFields(f);
  }
  throw new Error("handoff: expected a JSON object or a nachweis://handoff?... URI");
}

export function handoffUri(h: Handoff): string {
  const q = new URLSearchParams();
  q.set("v", "1");
  q.set("s", h.session_id);
  q.set("a", h.bound_address);
  q.set("c", h.challenge_hex);
  if (h.verifier_url) q.set("r", h.verifier_url);
  if (h.bridge_url) q.set("b", h.bridge_url);
  return `nachweis://handoff?${q.toString()}`;
}
